libs/@local/harpc/net/src/session/server/transaction/mod.rs

```diff
@@ -0,0 +1,415 @@
+    },
+    response::{kind::ResponseKind, Response},
+};
+use libp2p::PeerId;
+use tokio::{select, sync::mpsc};
+use tokio_util::{
+    sync::{CancellationToken, PollSendError, PollSender},
+    task::TaskTracker,
+};
+
+use super::{connection::collection::TransactionPermit, session_id::SessionId, SessionConfig};
+use crate::session::{
+    error::TransactionError,
+    writer::{ResponseContext, ResponseWriter, WriterOptions},
+};
+
+pub(crate) trait ServerTransactionPermit: Send + Sync + 'static {
+    fn id(&self) -> RequestId;
+    fn cancellation_token(&self) -> &CancellationToken;
+}
+
+struct TransactionSendDelegateTask<P> {
+    config: SessionConfig,
+
+    // TODO: consider switching to `tachyonix` crate for better performance (not yet tested)
+    // as well as more predictable buffering behavioud. `PollSender` is prone to just buffer
+    // everything before sending, which might not be the best idea in this scenario.
+    rx: mpsc::Receiver<core::result::Result<Bytes, TransactionError>>,
+    tx: mpsc::Sender<Response>,
+
+    permit: Arc<P>,
+}
+
+impl<P> TransactionSendDelegateTask<P>
+where
+    P: ServerTransactionPermit,
+{
+    #[expect(
+        clippy::integer_division_remainder_used,
+        reason = "required for select! macro"
+    )]
+    async fn run(mut self) {
+        let cancel = self.permit.cancellation_token();
+
+        // we cannot simply forward here, because we want to be able to send the end of request and
+        // buffer the response into the least amount of packages possible
+
+        let mut writer = ResponseWriter::new(
+            WriterOptions {
+                no_delay: self.config.no_delay,
+            },
+            ResponseContext {
+                id: self.permit.id(),
+                kind: ResponseKind::Ok,
+            },
+            &self.tx,
+        );
+
+        loop {
+            // TODO: potential performance improvement: use `recv_many` instead of `recv`
+            // This only is a performance improvement depending on the behaviour of the channel
+            // used. `PollSender` likes to buffer as much as possible before sending, which might
+            // not be a good idea.
+            let bytes = select! {
+                bytes = self.rx.recv() => bytes,
+                () = cancel.cancelled() => {
+                    break;
+                },
+            };
+
+            let Some(bytes) = bytes else {
+                // channel has been closed, we are done, flush the buffer
+                // flush the remaining buffer, if there's any
+                if let Err(error) = writer.flush().await {
+                    tracing::error!(?error, "connection has been prematurely closed");
+                }
+
+                break;
+            };
+
+            match bytes {
+                Ok(_) if writer.is_error() => {
+                    // we had an error previously, so just ignore the rest of the stream
+                    continue;
+                }
```

Comment:
Does the `writer` print the error?

Reply:
Yes, this is just in the case that we have a stream of:

```
Ok, Ok, Err, Ok, Err, Err
```

In that case, we can safely drop the `Ok` between the errors because we will error out anyway. `is_error` means that the current stream we're transmitting is an error stream.